cgnsutilities/cgnsutilities.py

```diff
@@ -1602,34 +1616,46 @@ def coarsen(self):
         conditions and B2B if necessary"""
         # We will coarsen one direction at a time. We do this to check if the block
         # is already 1-cell wide, which can't be coarsened any further
-        if self.dims[0] != 2:
-            self.coords = self.coords[0::2, :, :, :]
-            # Update BCs and B2B
-            for boco in self.bocos:
-                boco.coarsen(0)
-            for b2b in self.B2Bs:
-                b2b.coarsen(0)
 
-        if self.dims[1] != 2:
-            self.coords = self.coords[:, 0::2, :, :]
-            # Update BCs and B2B
-            for boco in self.bocos:
-                boco.coarsen(1)
-            for b2b in self.B2Bs:
-                b2b.coarsen(1)
+        # the new dimensions are half rounded up of the old dimensions
+        new_dims = copy.deepcopy(self.dims)
+        for i in range(3):
+            if self.dims[i] > 2:
+                new_dims[i] = (self.dims[i] + 1) // 2
+
+        new_coords = numpy.zeros((new_dims[0], new_dims[1], new_dims[2], 3))
+
+        # Loop over all directions
+        s = slice(None)
+        fine_slicer = [s] * 3
+        for idx_dim in range(3):
+
+            # can this direction be coarsened?
+            if new_dims[idx_dim] > 2:
+
+                fine_slicer[idx_dim] = slice(None, None, 2)
 
-        if self.dims[2] != 2:
-            self.coords = self.coords[:, :, 0::2, :]
-            # Update BCs and B2B
             for boco in self.bocos:
-                boco.coarsen(2)
+                boco.coarsen(idx_dim)
             for b2b in self.B2Bs:
-                b2b.coarsen(2)
+                b2b.coarsen(idx_dim)
 
-        # Update dimensions
-        self.dims[0] = self.coords.shape[0]
-        self.dims[1] = self.coords.shape[1]
-        self.dims[2] = self.coords.shape[2]
+        new_coords = self.coords[tuple(fine_slicer)]
+
+        # set the last ponint to be the same so we don't create any gaps if
+        # the number of points in that direction isn't odd
+        for idx_dim in range(3):
+            end_fine_slicer = copy.deepcopy(fine_slicer)
+            end_coarse_slicer = [s] * 3
```

Comment:
I don't really understand this part. It seems like the end slicers just need to be lists or tuples of dimension 3, not `slice` objects.

Reply:
You could probably do it without slicers, but I think this is cleaner. You would need to add if statements for each dimension.  
```
if idx_dim==0:
    fine_mapping[0] = -1
    new_coords[-1, :, :] = coords[fine_mapping]
elif idx_dim==1:    
    fine_mapping[1] = -1
    new_coords[:,-1, :] = coords[fine_mapping]
    
```

Reply:
This looks fine to me; I agree that the slicer approach is a bit cleaner so sure. 

It also looks like the edge cases of corners would also be handled properly, but I am not 100% sure I could imagine the issues that might come up. As long as you thought about the edge corners getting handled properly (e.g. coarsening on both i, j, and k with an even number of cells on all directions), then this is fine.